### PR TITLE
GH-84 swagger endpoint to receive block

### PR DESCRIPTION
**Pull request by velzevur**

### CODE REVIEW

File: apps/aecore/src/aec_chain.erl [hunk 1]
```diff
@@ -712,8 +712,6 @@ do_get_header_by_hash(HeaderHash, TopHeader, HeadersDb) ->
 -type do_get_block_by_hash_reply() ::
         {ok, header()} |
         {error, Reason::{block_not_found, {top_header, header()}}}.
--spec do_get_block_by_hash(block_header_hash(), chain_header(), blocks_db()) ->
```
Comment (lucafavatella) [inline apps/aecore/src/aec_chain.erl]: Spec lost - probably by mistake.

File: apps/aecore/src/aec_headers.erl [hunk 2]
```diff
@@ -38,24 +36,55 @@ time_in_secs(Header) ->
...
+    %  (H#header.nonce):64,
+    %  (H#header.time):64,
+    %  (H#header.version):16>>.
+
+-spec deserialize_from_binary(header_binary()) -> {ok, header()}.
+deserialize_from_binary(B) ->
+    deserialize_from_map(jsx:decode(B, [return_maps])).
+
+-spec hash_header(header()) -> {ok, block_header_hash()}.
+hash_header(H) ->
```
Comment (lucafavatella) [inline apps/aecore/src/aec_headers.erl]: Please add TODO comment mentioning that this hash is meant to be deterministic and based on the canonical serialization of the header.

File: apps/aecore/include/blocks.hrl [hunk 3]
```diff
@@ -30,3 +30,4 @@
           version = 0             :: non_neg_integer(),
           pow_evidence = no_value :: aec_pow:pow_evidence()}).
 -type(header() :: #header{}).
+-type(header_binary() :: binary()).
```
Comment (lucafavatella) [inline apps/aecore/include/blocks.hrl]: Please add comment clarifying that this is meant to be the deterministic canonical serialization of a header.

File: config/swagger.yaml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        '200':
+          description: top block
+#          schema:
+#            $ref: '#/definitions/Block'
+      security: []
+    post:
+      tags:
+        - external
+      operationId: PutBlock 
+      description: 'Get the top block header'
```
Comment (lucafavatella) [inline config/swagger.yaml]: Copy-paste error?

File: config/swagger.yaml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      parameters: []
+      responses:
+        '200':
+          description: top block
+#          schema:
+#            $ref: '#/definitions/Block'
```
Comment (lucafavatella) [inline config/swagger.yaml]: I see `Block` is defined and used below. Any reason why commented?

File: config/swagger.yaml [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  Tree:
+    type: object
+    properties: {}
+  Block:
+    type: object
+    properties:
+      txs:
+        type: array
+        items:
+          $ref: '#/definitions/Tx'
```
Comment (lucafavatella) [inline config/swagger.yaml]: I do not see`Tx` defined. I guess swagger generator does not complain.

Comment (lucafavatella) [reply config/swagger.yaml]: I comment in the PR that this is TODO would be nice to have.

File: config/swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      trees:
+        type: array
+        items:
+          $ref: '#/definitions/Tree'
+      height:
+        type: integer
+        format: int64
+      prev-hash:
+        type: string
+      root-hash:
```
Comment (lucafavatella) [inline config/swagger.yaml]: In the further iteration of this schema I suggest we prefer names from the canonical description https://github.com/aeternity/epoch/wiki/Block-Header e.g. state-hash

File: apps/aecore/src/aec_mining.erl [hunk 8]
```diff
@@ -45,7 +45,7 @@ mine(Attempts) ->
 -spec mine(block(), non_neg_integer()) -> {ok, block()} | {error, term()}.
 mine(Block, Attempts) ->
     Target = aec_blocks:target(Block),
-    {ok, BlockBin} = aec_headers:serialize_for_network(aec_blocks:to_header(Block)),
+    {ok, BlockBin} = aec_headers:serialize_to_map(aec_blocks:to_header(Block)),
```
Comment (uwiger) [inline apps/aecore/src/aec_mining.erl]: Perhaps change that variable name BlockBin, given that `serialize_to_map()` returns a map?

File: config/swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+          description: successful operation
+      security: []
+definitions:
+  Tree:
+    type: object
+    properties: {}
+  Block:
+    type: object
+    properties:
+      trees:
```
Comment (lucafavatella) [inline config/swagger.yaml]: The trees must not be serialized.

**Comment by velzevur** [comment]
superseded by #154

**Pull request closed without merging**